### PR TITLE
add extensions logic and mysql-extension

### PR DESCRIPTION
**Pull request by g0djan**

### CODE REVIEW

File: extensions/mysql/lib/mysql.go [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+		if err != nil {
+			tracelog.ErrorLogger.Fatalf("%+v\n", err)
+		}
+		return logFileName
+	}
+	rows, err = db.Query("SHOW SLAVE STATUS")
+	if err != nil {
+		tracelog.ErrorLogger.Fatalf("%+v\n", err)
+	}
+	for rows.Next() {
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: Also, don't need a 'for' loop here.

File: extensions/mysql/lib/mysql.go [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+			tracelog.ErrorLogger.Fatalf("%+v\n", err)
+		}
+		return logFileName
+	}
+	tracelog.ErrorLogger.Fatalf("Failed to obtain current binlog file")
+	return ""
+}
+
+func getMySQLConnection() (*sql.DB, error) {
+	datasourceName := internal.GetSettingValue("WALG_MYSQL_DATASOURCE_NAME")
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: What about adding a default value parameter to GetSettingValue function?

File: extensions/mysql/lib/mysql.go [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+	if datasourceName == "" {
+		datasourceName = "root:password@/mysql"
+	}
+	caFile := internal.GetSettingValue("WALG_MYSQL_SSL_CA")
+	if caFile != "" {
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: It should be moved to function like "AddSSLToDatasource".

File: extensions/mysql/lib/mysql.go [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+func uploadStreamSentinel(sentinelDto *StreamSentinelDto, uploader *Uploader, name string) error {
+	dtoBody, err := json.Marshal(*sentinelDto)
+	if err != nil {
+		return err
+	}
+
+	uploadingErr := uploader.Upload(name, bytes.NewReader(dtoBody))
+	if uploadingErr != nil {
+		tracelog.ErrorLogger.Printf("upload: could not upload '%s'\n", name)
+		tracelog.ErrorLogger.Fatalf("StorageTarBall finish: json failed to upload")
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: You have to either fatal or return here, but not both. I strongly recommend you using return.

File: extensions/mysql/lib/mysql.go [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+		return uploadingErr
+	}
+	return nil
+}
+
+func getStreamName(backup *Backup, extension string) string {
+	dstPath := internal.SanitizePath(backup.Name+"/stream.") + extension
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: Let's join paths with Golang internal function instead of simple '+'.

File: extensions/mysql/lib/mysql.go [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+		if err != nil {
+			return nil, err
+		}
+		if ok := rootCertPool.AppendCertsFromPEM(pem); !ok {
+			return nil, fmt.Errorf("Failed to load certificate from %s", caFile)
+		}
+		mysql.RegisterTLSConfig("custom", &tls.Config{
+			RootCAs: rootCertPool,
+		})
+		if strings.Contains(datasourceName, "?tls=") || strings.Contains(datasourceName, "&tls=") {
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: Let's replace it with methods:
ContainsOption()
AddOption()

File: extensions/mysql/lib/mysql.go [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+		tracelog.ErrorLogger.Fatalf("%+v\n", err)
+	}
+	for rows.Next() {
+		err = rows.Scan(&logFileName, &garbage, &garbage, &garbage, &garbage)
+		if err != nil {
+			tracelog.ErrorLogger.Fatalf("%+v\n", err)
+		}
+		return logFileName
+	}
+	tracelog.ErrorLogger.Fatalf("Failed to obtain current binlog file")
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: Let's remove 'fatal' from here and return error as it is.

File: extensions/mysql/lib/mysql.go [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+		var size uint32
+		err = rows.Scan(&logFinName, &size)
+		if err != nil {
+			tracelog.ErrorLogger.Fatalf("%+v\n", err)
+		}
+		if logFinName < currentBinlog {
+			result = append(result, logFinName)
+		}
+	}
+	sort.Strings(result)
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: Go is so sick...

File: extensions/mysql/lib/mysql.go [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+
+	if backupName == "" {
+		backupName = StreamPrefix + time.Now().UTC().Format("20060102T150405Z")
+	}
+	stat, _ := os.Stdin.Stat()
+	var stream io.Reader = os.Stdin
+	if (stat.Mode() & os.ModeCharDevice) == 0 {
+		tracelog.InfoLogger.Println("Data is piped from stdin")
+		//tracelog.ErrorLogger.Println("WARNING: operating in test mode!")
+		//stream = strings.NewReader("testtesttest")
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: Let's clear code here :D

File: extensions/mysql/lib/mysql.go [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+				binlogsAreDone <- err
+				return
+			}
+		}
+	}
+
+	binlogsAreDone <- nil
+}
+
+func CheckOldBinlogs(backupTime internal.BackupTime, folder internal.StorageFolder) {
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: This function not only Checks old binlogs, but also deletes them...

File: extensions/mysql/lib/mysql.go [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+		d := decompressor
+		archiveReader, exists, err := internal.TryDownloadWALFile(baseBackupFolder, getStreamName(&backup, d.FileExtension()))
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+		err = internal.DecompressWALFile(&internal.EmptyWriteIgnorer{WriteCloser: os.Stdout}, archiveReader, d)
+		os.Stdout.Close()
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: We are trying to close stdout in cycle - WTF?

File: extensions/mysql/lib/mysql.go [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+// TODO : unit tests
+func downloadAndDecompressStream(folder internal.StorageFolder, fileName string) error {
+	baseBackupFolder := folder.GetSubFolder(internal.BaseBackupPath)
+	backup := Backup{internal.NewBackup(baseBackupFolder, fileName)}
+
+	streamSentinel, err := backup.fetchStreamSentinel()
+	if err != nil {
+		return err
+	}
+	binlogsAreDone := make(chan error)
```
Comment (Tinsane) [inline extensions/mysql/lib/mysql.go]: It doesn't show us, whether binlogs are done or not. It shows us, whether we had an error during binlogs fetching process.

File: cmd/wal-g/main.go [hunk 13]
```diff
@@ -9,8 +9,21 @@ import (
 	"runtime/pprof"
 )
 
+
+type arrayFlags []string
```
[the remark this answers is not in the document]
Comment (g0djan) [reply cmd/wal-g/main.go]: I simplify it, removed logic with cli args and without config. Now it just loads all ".so" files in "extensions" directory.

**Comment by g0djan** [comment]
I divide it into 2 simple PR.
1. Logic of extensions
2. MySQL extension

File: internal/storage_tar_ball.go [hunk 14]
```diff
@@ -133,18 +133,18 @@ func (tarBall *StorageTarBall) Finish(sentinelDto *BackupSentinelDto) error {
 	uploader.finish()
 
 	var err error
-	// If other parts are successful in uploading, upload json file.
+	// If other parts are successful in uploading, Upload json file.
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/storage_tar_ball.go]: Fixed

File: internal/uploader.go [hunk 15]
```diff
@@ -45,15 +45,15 @@ func NewUploader(
 func (uploader *Uploader) finish() {
 	uploader.waitGroup.Wait()
 	if !uploader.Success {
-		tracelog.ErrorLogger.Printf("WAL-G could not complete upload.\n")
+		tracelog.ErrorLogger.Printf("WAL-G could not complete Upload.\n")
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/uploader.go]: Fixed

File: internal/storage_tar_ball.go [hunk 16]
```diff
@@ -133,18 +133,18 @@ func (tarBall *StorageTarBall) Finish(sentinelDto *BackupSentinelDto) error {
...
 			return err
 		}
 
-		uploadingErr := uploader.upload(name, bytes.NewReader(dtoBody))
+		uploadingErr := uploader.Upload(name, bytes.NewReader(dtoBody))
 		if uploadingErr != nil {
-			tracelog.ErrorLogger.Printf("upload: could not upload '%s'\n", name)
-			tracelog.ErrorLogger.Fatalf("StorageTarBall finish: json failed to upload")
+			tracelog.ErrorLogger.Printf("Upload: could not Upload '%s'\n", name)
+			tracelog.ErrorLogger.Fatalf("StorageTarBall finish: json failed to Upload")
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/storage_tar_ball.go]: Fixed

File: internal/storage_tar_ball.go [hunk 17]
```diff
@@ -90,12 +90,12 @@ func (tarBall *StorageTarBall) startUpload(name string, crypter Crypter) io.Writ
...
 
-		err := uploader.upload(path, NewNetworkLimitReader(pipeReader))
+		err := uploader.Upload(path, NewNetworkLimitReader(pipeReader))
 		if compressingError, ok := err.(CompressingPipeWriterError); ok {
-			tracelog.ErrorLogger.Printf("could not upload '%s' due to compression error\n%+v\n", path, compressingError)
+			tracelog.ErrorLogger.Printf("could not Upload '%s' due to compression error\n%+v\n", path, compressingError)
 		}
 		if err != nil {
-			tracelog.ErrorLogger.Printf("upload: could not upload '%s'\n", path)
+			tracelog.ErrorLogger.Printf("Upload: could not Upload '%s'\n", path)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/storage_tar_ball.go]: Fixed

File: internal/storage_tar_ball.go [hunk 18]
```diff
@@ -90,12 +90,12 @@ func (tarBall *StorageTarBall) startUpload(name string, crypter Crypter) io.Writ
 	go func() {
 		defer uploader.waitGroup.Done()
 
-		err := uploader.upload(path, NewNetworkLimitReader(pipeReader))
+		err := uploader.Upload(path, NewNetworkLimitReader(pipeReader))
 		if compressingError, ok := err.(CompressingPipeWriterError); ok {
-			tracelog.ErrorLogger.Printf("could not upload '%s' due to compression error\n%+v\n", path, compressingError)
+			tracelog.ErrorLogger.Printf("could not Upload '%s' due to compression error\n%+v\n", path, compressingError)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/storage_tar_ball.go]: Fixed

File: internal/storage_tar_ball.go [hunk 19]
```diff
@@ -76,7 +76,7 @@ func (tarBall *StorageTarBall) AwaitUploads() {
 }
 
 // TODO : unit tests
-// startUpload creates a compressing writer and runs upload in the background once
+// startUpload creates a compressing writer and runs Upload in the background once
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/storage_tar_ball.go]: Fixed

File: internal/storage_tar_ball.go [hunk 20]
```diff
@@ -33,7 +33,7 @@ type StorageTarBall struct {
 	uploader    *Uploader
 }
 
-// SetUp creates a new tar writer and starts upload to storage.
+// SetUp creates a new tar writer and starts Upload to storage.
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/storage_tar_ball.go]: Fixed

File: internal/s3_uploader.go [hunk 21]
```diff
@@ -43,5 +43,5 @@ func (uploader *S3Uploader) createUploadInput(bucket, path string, content io.Re
 func (uploader *S3Uploader) upload(bucket, path string, content io.Reader) error {
 	input := uploader.createUploadInput(bucket, path, content)
 	_, err := uploader.uploaderAPI.Upload(input)
-	return errors.Wrapf(err, "failed to upload '%s' to bucket '%s'", path, bucket)
+	return errors.Wrapf(err, "failed to Upload '%s' to bucket '%s'", path, bucket)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/s3_uploader.go]: Fixed

File: internal/prefetch.go [hunk 22]
```diff
@@ -66,14 +66,14 @@ func prefaultData(prefaultStartLsn uint64, timelineId uint32, waitGroup *sync.Wa
...
-	err := bundle.DownloadDeltaMap(uploader.uploadingFolder.GetSubFolder(WalPath), prefaultStartLsn+WalSegmentSize*WalFileInDelta)
+	err := bundle.DownloadDeltaMap(uploader.UploadingFolder.GetSubFolder(WalPath), prefaultStartLsn+WalSegmentSize*WalFileInDelta)
 	if err != nil {
 		tracelog.ErrorLogger.Printf("Error during loading delta map: '%+v'.", err)
 		return
 	}
 	bundle.TarBallMaker = NewNopTarBallMaker()
 
-	// Start a new tar bundle, walk the archiveDirectory and upload everything there.
+	// Start a new tar bundle, walk the archiveDirectory and Upload everything there.
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/prefetch.go]: Fixed

File: internal/open_pgp_crypter.go [hunk 23]
```diff
@@ -79,7 +79,7 @@ func (crypter *OpenPGPCrypter) ConfigurePGPCrypter() {
 		return
 	}
 
-	// key can be either private (for download) or public (for upload)
+	// key can be either private (for download) or public (for Upload)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/open_pgp_crypter.go]: Fixed

File: internal/open_pgp_crypter.go [hunk 24]
```diff
@@ -69,7 +69,7 @@ func (crypter *OpenPGPCrypter) IsUsed() bool {
 func (crypter *OpenPGPCrypter) ConfigurePGPCrypter() {
 	crypter.Configured = true
 
-	// key can be either private (for download) or public (for upload)
+	// key can be either private (for download) or public (for Upload)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/open_pgp_crypter.go]: Fixed

File: internal/delta_file_manager.go [hunk 25]
```diff
@@ -169,11 +169,11 @@ func (manager *DeltaFileManager) FlushDeltaFiles(uploader *Uploader, completedPa
 			var deltaFileData bytes.Buffer
 			err := deltaFileWriter.DeltaFile.Save(&deltaFileData)
 			if err != nil {
-				tracelog.WarningLogger.Printf("Failed to upload delta file: '%s' because of saving error: '%v'\n", deltaFilename, err)
+				tracelog.WarningLogger.Printf("Failed to Upload delta file: '%s' because of saving error: '%v'\n", deltaFilename, err)
 			} else {
 				err = uploader.UploadFile(&NamedReaderImpl{&deltaFileData, deltaFilename})
 				if err != nil {
-					tracelog.WarningLogger.Printf("Failed to upload delta file: '%s' because of uploading error: '%v'\n", deltaFilename, err)
+					tracelog.WarningLogger.Printf("Failed to Upload delta file: '%s' because of uploading error: '%v'\n", deltaFilename, err)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/delta_file_manager.go]: Fixed

File: internal/delta_file_manager.go [hunk 26]
```diff
@@ -169,11 +169,11 @@ func (manager *DeltaFileManager) FlushDeltaFiles(uploader *Uploader, completedPa
 			var deltaFileData bytes.Buffer
 			err := deltaFileWriter.DeltaFile.Save(&deltaFileData)
 			if err != nil {
-				tracelog.WarningLogger.Printf("Failed to upload delta file: '%s' because of saving error: '%v'\n", deltaFilename, err)
+				tracelog.WarningLogger.Printf("Failed to Upload delta file: '%s' because of saving error: '%v'\n", deltaFilename, err)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/delta_file_manager.go]: Fixed

File: internal/configure.go [hunk 27]
```diff
@@ -292,7 +292,7 @@ func configureLogging() error {
 // Requires these environment variables to be set:
 // WALE_S3_PREFIX
 //
-// Able to configure the upload part size in the S3 uploader.
+// Able to configure the Upload part size in the S3 uploader.
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/configure.go]: Fixed

File: internal/configure.go [hunk 28]
```diff
@@ -38,7 +38,7 @@ func (err SseKmsIdNotSetError) Error() string {
 	return fmt.Sprintf(tracelog.GetErrorFormatter(), err.error)
 }
 
-// MaxRetries limit upload and download retries during interaction with S3
+// MaxRetries limit Upload and download retries during interaction with S3
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/configure.go]: Fixed

File: internal/compressing_pipe_writer.go [hunk 29]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 // CompressingPipeWriterError is used to catch specific errors from CompressingPipeWriter
-// when uploading to S3. Will not retry upload if this error
+// when uploading to S3. Will not retry Upload if this error
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/compressing_pipe_writer.go]: Fixed

File: internal/bguploader.go [hunk 30]
```diff
@@ -13,7 +13,7 @@ import (
 
 const TotalBgUploadedLimit = 1024
 
-// BgUploader represents the state of concurrent WAL upload
+// BgUploader represents the state of concurrent WAL Upload
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/bguploader.go]: Fixed

File: internal/backup_push_handler.go [hunk 31]
```diff
@@ -123,11 +123,11 @@ func HandleBackupPush(archiveDirectory string, uploader *Uploader) {
 	if err != nil {
 		tracelog.ErrorLogger.FatalError(err)
 	}
-	err = bundle.UploadPgControl(uploader.compressor.FileExtension())
+	err = bundle.UploadPgControl(uploader.Compressor.FileExtension())
 	if err != nil {
 		tracelog.ErrorLogger.FatalError(err)
 	}
-	// Stops backup and write/upload postgres `backup_label` and `tablespace_map` Files
+	// Stops backup and write/Upload postgres `backup_label` and `tablespace_map` Files
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/backup_push_handler.go]: Fixed

File: internal/backup_push_handler.go [hunk 32]
```diff
@@ -112,7 +112,7 @@ func HandleBackupPush(archiveDirectory string, uploader *Uploader) {
 
 	bundle.TarBallMaker = NewStorageTarBallMaker(backupName, uploader)
 
-	// Start a new tar bundle, walk the archiveDirectory and upload everything there.
+	// Start a new tar bundle, walk the archiveDirectory and Upload everything there.
```
[the remark this answers is not in the document]
Comment (g0djan) [reply internal/backup_push_handler.go]: Fixed

File: extensions/mysql/README.md [hunk 33]
```diff
@@ -0,0 +1,18 @@
+How to include MySQL extension to project
+-----
+1. You should to compile folder mysql with options
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/README.md]: Fixed

File: cmd/wal-g/main.go [hunk 34]
```diff
@@ -137,6 +161,8 @@ func main() {
 		internal.HandleBackupList(folder)
 	} else if command == "delete" {
 		internal.HandleDelete(folder, all)
+	} else if extension, ok := internal.GetExtensionByCommandName(command); ok{
```
[the remark this answers is not in the document]
Comment (g0djan) [reply cmd/wal-g/main.go]: Done

File: extensions/mysql/extension.go [hunk 35]
```diff
@@ -0,0 +1,61 @@
+package main
+
+import (
+	"fmt"
+	"github.com/wal-g/wal-g/extensions/mysql/lib"
+	"github.com/wal-g/wal-g/internal"
+)
+
+type extension struct {
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/extension.go]: Done

File: extensions/mysql/lib/mysql.go [hunk 36]
```diff
@@ -0,0 +1,439 @@
+package lib
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go [hunk 37]
```diff
@@ -0,0 +1,439 @@
...
+)
+
+type Uploader struct {
+	*internal.Uploader
+}
+type Backup struct {
+	*internal.Backup
+}
+
+type ArchiveNonExistenceError internal.ArchiveNonExistenceError
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Removed

File: extensions/mysql/lib/mysql.go (hunk 36 above)
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go [hunk 38]
```diff
@@ -0,0 +1,439 @@
...
+		NewCompressingWriter: compressor.NewWriter,
+	}
+
+	pipeWriter.Compress(&internal.OpenPGPCrypter{})
+
+	backup := Backup{internal.NewBackup(uploader.UploadingFolder, fileName)}
+
+	dstPath := getStreamName(&backup, compressor.FileExtension())
+	tracelog.DebugLogger.Println("Upload path", dstPath)
+	reader := pipeWriter.Output
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go [hunk 39]
```diff
@@ -0,0 +1,439 @@
...
+const (
+	BinlogPath               = "binlog_" + internal.VersionStr + "/"
+	MysqlBinlogCacheFileName = ".walg_mysql_logs_cache"
+	StreamPrefix             = "stream_"
+)
+
+func HandleMySQLCron(uploader *Uploader, backupLocation string) {
+	uploader.UploadingFolder = uploader.UploadingFolder.GetSubFolder(BinlogPath)
+	db, err := getMySQLConnection()
+	defer db.Close()
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Fixed

File: extensions/mysql/lib/mysql.go [hunk 40]
```diff
@@ -0,0 +1,439 @@
...
+	if log <= getLastArchivedBinlog() {
+		tracelog.InfoLogger.Printf("Binlog %v already archived\n", log)
+		return nil
+	}
+	tracelog.InfoLogger.Printf("Archiving %v\n", log)
+
+	walFile, err := os.Open(filename)
+	if err != nil {
+		return errors.Wrapf(err, "upload: could not open '%s'\n", filename)
+	}
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Fixed

File: extensions/mysql/lib/mysql.go [hunk 41]
```diff
@@ -0,0 +1,439 @@
...
+		tracelog.ErrorLogger.Println("WARNING: stdin is terminal: operating in test mode!")
+		stream = strings.NewReader("testtesttest")
+	}
+	err = uploader.UploadStream(backupName, db, stream)
+	if err != nil {
+		tracelog.ErrorLogger.Fatalf("%+v\n", err)
+	}
+}
+
+const (
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go (hunk 5 above)
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go [hunk 42]
```diff
@@ -0,0 +1,439 @@
...
+		cacheFilename = filepath.Join(usr.HomeDir, MysqlBinlogCacheFileName)
+		var file []byte
+		file, err = ioutil.ReadFile(cacheFilename)
+		// here we ignore whatever error can occur
+		if err == nil {
+			_ = json.Unmarshal(file, &cache)
+		}
+	}
+	if err != nil && !os.IsNotExist(err) {
+		tracelog.ErrorLogger.Printf("%+v\n", err)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Replaced with  "tracelog.ErrorLogger.Fatal"

File: extensions/mysql/lib/mysql.go [hunk 43]
```diff
@@ -0,0 +1,439 @@
...
+	return ""
+}
+
+func setLastArchivedBinlog(binlogFileName string) {
+	var cache MySQLLogsCache
+	var cacheFilename string
+
+	usr, err := user.Current()
+	if err == nil {
+		cacheFilename = filepath.Join(usr.HomeDir, MysqlBinlogCacheFileName)
```
[the remark this answers is not in the document]
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Replaced with filepath.Join("tmp", usr.Name, MysqlBinlogCacheFileName)

File: extensions/mysql/lib/mysql.go (hunk 2 above)
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go (hunk 5 above)
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Done

File: extensions/mysql/lib/mysql.go (hunk 10 above)
Comment (g0djan) [reply extensions/mysql/lib/mysql.go]: Renamed to DeleteOldBinlogs, because it seems to be ok if it wouldn't be found

**Pull request closed without merging**